packages/react-server/src/ReactFizzServer.js

```diff
@@ -159,6 +159,14 @@ const ERRORED = 4;
 
 type Root = null;
 
+const TEXT_EDGE = 0;
+const NODE_EDGE = 1;
+
+// A SegmentEdge is a descriptor for the kind of object on either side of a segments boundary (leading and trailing).
+// In some cases the Edge type isn't known because the edge is a child segment itself. For these cases we use
+// the child segment to identify as the edge unless and until a known edge type is found
+type SegmentEdge = 0 | 1 | Segment;
```

Comment:
We shouldn't mix number and pointer types. This will mess up the JIT. Always use types as you would in Rust. (This means that we should really not mix `null` and numbers neither.)

Storing the pointers is not great for memory management purposes when we port this. In general I try to avoid cross-linking between objects unless absolutely necessary since it breaks the simplicity of ownership. It also makes GC less predictable.

It also makes it harder to do tricks like moving segments around or reusing cached ones if the structure assumes some kind of graph.